### ISSUE TITLE
dataset

### ISSUE THREAD

**Issue by Zhenya97** [CaelynCheung1996/QuickNAT_tensorflow_V2#1]
Hi Caelyn,

Thank you for your work. Could you please share the dataset that you used with me? I appreciate it if you can send a copy to me. My email address is chris4771zz@gmail.com.

Thanks,
Sincerely,
Chris

**Comment by Zhenya97** [comment]
Could you please give more explanation about this:

train_data = train_data.reshape(train_data.shape[0], train_data.shape[2], train_data.shape[3], train_data.shape[1]) # (NumOfData, height, weight, channel=1 for grayscale data)

Thanks

**Comment by CaelynCheung1996** [comment]
> Hi Caelyn,
> 
> Thank you for your work. Could you please share the dataset that you used with me? I appreciate it if you can send a copy to me. My email address is [chris4771zz@gmail.com](mailto:chris4771zz@gmail.com).
> 
> Thanks,
> Sincerely,
> Chris

Hi Chris,
I didn't have the data with me in hand right now but I did put the link to download the dataset in README file of the repository. Please feel free to fill the form and you will get the link to download the dataset (you can just fill any team name that you come up with since it doesn't really relate to the challenge anymore, which should be free and available to download). 

Thanks,
Caelyn

**Comment by CaelynCheung1996** [comment]
> Could you please give more explanation about this:
> 
> train_data = train_data.reshape(train_data.shape[0], train_data.shape[2], train_data.shape[3], train_data.shape[1]) # (NumOfData, height, weight, channel=1 for grayscale data)
> 
> Thanks

This is just how TensorFlow handles the dataset and store them in a tensor. As long as you convert your own dataset into imdb.mat using the code that I provided, you can specify the path to that .mat file and the code should work. Thanks.

**Comment by Zhenya97** [comment]
Thank you very much. I have two more problems about the codes.

1. How to apply the code to 192 *256 *192 image dataset? 
   What parts of the codes need to be changed?
2.How to include multiple nii.gz data files in one imdbTraining.mat file?

I appreciate it if you can give some help with these problems.

Thanks,
Sincerely,
Chris

**Comment by CaelynCheung1996** [comment]
1. You can change the variable nVx, nVy to define the width and height of the image. And NumOfSlice_images, NumOfSlice_labels to define the number of slices in each volumetric image.
`nVx = 256`
`nVy = 256`
`NumOfSlice_images = 256 `
`NumOfSlice_labels = 256`
2. The code already be able to run through each image file in the specified folder where all the image data are stored. You can define to paths to images and labels in filelist_images and filelist_labels.

**Comment by CaelynCheung1996** [comment]
> Thank you very much. I have two more problems about the codes.
> 
> 1. How to apply the code to 192 *256 *192 image dataset?
>    What parts of the codes need to be changed?
>    2.How to include multiple nii.gz data files in one imdbTraining.mat file?
> 
> I appreciate it if you can give some help with these problems.
> 
> Thanks,
> Sincerely,
> Chris

1. You can change the variable nVx, nVy to define the width and height of the image. And NumOfSlice_images, NumOfSlice_labels to define the number of slices in each volumetric image.
`nVx = 256`
`nVy = 256`
`NumOfSlice_images = 256`
`NumOfSlice_labels = 256`
2. The code is capable to run through each image file in the specified folder where all the image data are stored. You can define to paths to images and labels in filelist_images and filelist_labels.

**Comment by Zhenya97** [comment]
Thank you very much. Thank you for your patience.
Now, the codes works well with 256*256*256 images dataset.
Then, what should I change with the 'training.py' and 'testing.py' to apply them to 192*256*192 image datasets?
I appreciate it if you can give some directions.
Please also forgive my ignorance. This is the first time I study relevant knowledge.

Thanks,
Sincerely,
Chris

**Comment by CaelynCheung1996** [comment]
> Thank you very much. Thank you for your patience.
> Now, the codes works well with 256_256_256 images dataset.
> Then, what should I change with the 'training.py' and 'testing.py' to apply them to 192_256_192 image datasets?
> I appreciate it if you can give some directions.
> Please also forgive my ignorance. This is the first time I study relevant knowledge.
> 
> Thanks,
> Sincerely,
> Chris

The values of 256 in those script mostly correspond to the height and width of the images of my dataset. You can read through the script and change the values to dimension of you own ones. I guess those are the only values you need to change in these scripts.

**Comment by Zhenya97** [comment]
Thank you very much. Now the codes work well. 
However, it runs very slowly and seems to be on the CPU instead of the GPU. My computer configuration is a six-core processor and a GTX 1070oc 8G GPU. The 'imdbTraining' file I used contains only three data sets. Is it something I have neglected? I appreciate it if you can give some advise to reduce the running time.

Thanks,
Sincerely,
Chris

**Comment by Zhenya97** [comment]
Should I reduce the times of epochs? or something else

**Comment by CaelynCheung1996** [comment]
> Thank you very much. Now the codes work well.
> However, it runs very slowly and seems to be on the CPU instead of the GPU. My computer configuration is a six-core processor and a GTX 1070oc 8G GPU. The 'imdbTraining' file I used contains only three data sets. Is it something I have neglected? I appreciate it if you can give some advise to reduce the running time.
> 
> Thanks,
> Sincerely,
> Chris

If you are not using singularity containers as described in my README file, you should be able to install tensorflow-gpu via conda. Before you run the code, you should "export CUDA_VISIBLE_DEVICES = 0" (or other GPU IDs). Since I run this quite a while ago, it may be also relevant to the RAM limitation of your GPU. The number of epoch is also one of the hyper-parameters to tune so feel free to play with that until you get satisfying results.

**Comment by CaelynCheung1996** [comment]
> Should I reduce the times of epochs? or something else

Reducing batch size may give it a shot.

**Comment by Zhenya97** [comment]
Thank you very much. Singularity containers? You mean preprocess of the datasets by 'caculate_weights.py' and 'remapping_labels.py'?

**Comment by CaelynCheung1996** [comment]
> Thank you very much. Singularity containers? You mean preprocess of the datasets by 'caculate_weights.py' and 'remapping_labels.py'?

No. That's just the environment containing tensorflow. You can install tensorflow in your own way.

**Comment by Zhenya97** [comment]
Thank you very much. Appreciate it.

**Comment by Zhenya97** [comment]
I still have a  question. I appreciate it if you have some time to give me some advise.
If I want to use the model trained by the training.py file to segment the images in a image dataset without the any label file (for example, a raw data file like AAAA.nii.gz) and output the segmentation result as an image file with any format. What should I modify the testing.py file? Should I delete the part of 'test_label' ?

Thanks,
Chris

**Comment by CaelynCheung1996** [comment]
> I still have a question. I appreciate it if you have some time to give me some advise.
> If I want to use the model trained by the training.py file to segment the images in a image dataset without the any label file (for example, a raw data file like AAAA.nii.gz) and output the segmentation result as an image file with any format. What should I modify the testing.py file? Should I delete the part of 'test_label' ?
> 
> Thanks,
> Chris

I modified the function of read_dataset, which should return only dataset only (no labels) right now. I also correct the corresponding line in test.py. Feel free to try and the testing.py is short to figure out. I think it will be a good practice to understanding these lines. Let me know if you have further questions.

**Comment by Zhenya97** [comment]
Thank you very much.

**Comment by Zhenya97** [comment]
I still have a few small doubts. 
1. What is vdc? 
2. What is ‘outfile.jpg'? The outfile.jpg I got is just a black background. And the imsave statement seems to no longer work in the new version. ('DeprecationWarning: `imsave` is deprecated!
`imsave` is deprecated in SciPy 1.0.0, and will be removed in 1.2.0.
Use ``imageio.imwrite`` instead.')
3. Where can I see the accuracy of the model, such as the dice score.

Thanks,
Chris

**Comment by CaelynCheung1996** [comment]
> I still have a few small doubts.
> 
> 1. What is vdc?
> 2. What is ‘outfile.jpg'? The outfile.jpg I got is just a black background. And the imsave statement seems to no longer work in the new version. ('DeprecationWarning: `imsave` is deprecated!
>    `imsave` is deprecated in SciPy 1.0.0, and will be removed in 1.2.0.
>    Use `imageio.imwrite` instead.')
> 3. Where can I see the accuracy of the model, such as the dice score.
> 
> Thanks,
> Chris

1. vdc is just a label variable I use for my own dataset.
2. It happens a lot for a function is updated. Just try what the system suggested and explore. ```imageio.imwrite``` is commonly used. I'm not sure why the output is all zeros but it shouldn't be if you have a trained model and apply it to test data.
3. The output for the dice score is only calculated in the training process, which you should be able to see when you train the model. You may need to change the code to incorporate calculating the dice score when you are testing the model.

**Comment by Zhenya97** [comment]
Thank you very much.

Could you please give more explanation about this part of the codes?

def split_dataset(labels, dataset="liver_spleen"):
    if dataset == "liver_spleen":
        label_liver_spleen = np.copy(np.array(labels))
        for label in label_liver_spleen:
            print(label.shape)
            # indexed array
            for i in range(4, 12):
                label[0, :, :][label[0, :, :] == i] = 1
        return label_liver_spleen[:, 0, :, :]


    if dataset == "rlung_llung":
        label_rlung_llung = np.copy(np.array(labels))
        for label in label_rlung_llung:
            # indexed array
            for i in range(8, 12):
                label[0, :, :][label[0, :, :] == i] = 1
            for j in range(2, 6):
                label[0, :, :][label[0, :, :] == j] = 1
        return label_rlung_llung[:, 0, :, :]

    if dataset == "rkidney_lkidney":
        label_rkidney_lkidney = np.copy(np.array(labels))
        for label in label_rkidney_lkidney:
            # indexed array
            for i in range(10, 12):
                label[0, :, :][label[0, :, :] == i] = 1
            for j in range(2, 8):
                label[0, :, :][label[0, :, :] == j] = 1
        return label_rkidney_lkidney[:, 0, :, :]


    if dataset == "liver_kidney_lung":
        liver_kidney_lung = np.copy(np.array(labels))
        for label in liver_kidney_lung:
            for i in range(10, 12):
                label[0, :, :][label[0, :, :] == i] = 1
            for j in range(4, 6):
                label[0, :, :][label[0, :, :] == j] = 1
        return liver_kidney_lung[:, 0, :, :]

Thanks,
Chris

**Comment by Zhenya97** [comment]
Is there any way to view  dice score of the model after training complete?

**Comment by Zhenya97** [comment]
Hi Caelyn,

I have tried your codes with lots of datasets. But the result is still empty. It really confused me. Could you please apply your method with some sample datasets for example? If you want some datasets, I can send them to you.

Thanks,
Sincerely,
Xinze

**Comment by CaelynCheung1996** [comment]
> Thank you very much.
> 
> Could you please give more explanation about this part of the codes?
> 
> def split_dataset(labels, dataset="liver_spleen"):
> if dataset == "liver_spleen":
> label_liver_spleen = np.copy(np.array(labels))
> for label in label_liver_spleen:
> print(label.shape)
> # indexed array
> for i in range(4, 12):
> label[0, :, :][label[0, :, :] == i] = 1
> return label_liver_spleen[:, 0, :, :]
> 
> ```
> if dataset == "rlung_llung":
>     label_rlung_llung = np.copy(np.array(labels))
>     for label in label_rlung_llung:
>         # indexed array
>         for i in range(8, 12):
>             label[0, :, :][label[0, :, :] == i] = 1
>         for j in range(2, 6):
>             label[0, :, :][label[0, :, :] == j] = 1
>     return label_rlung_llung[:, 0, :, :]
> 
> if dataset == "rkidney_lkidney":
>     label_rkidney_lkidney = np.copy(np.array(labels))
>     for label in label_rkidney_lkidney:
>         # indexed array
>         for i in range(10, 12):
>             label[0, :, :][label[0, :, :] == i] = 1
>         for j in range(2, 8):
>             label[0, :, :][label[0, :, :] == j] = 1
>     return label_rkidney_lkidney[:, 0, :, :]
> 
> 
> if dataset == "liver_kidney_lung":
>     liver_kidney_lung = np.copy(np.array(labels))
>     for label in liver_kidney_lung:
>         for i in range(10, 12):
>             label[0, :, :][label[0, :, :] == i] = 1
>         for j in range(4, 6):
>             label[0, :, :][label[0, :, :] == j] = 1
>     return liver_kidney_lung[:, 0, :, :]
> ```
> 
> Thanks,
> Chris

I don't think this function is used in the training.py. You can see that the function ```from sklearn.model_selection import train_test_split``` is used instead.

**Comment by CaelynCheung1996** [comment]
> Hi Caelyn,
> 
> I have tried your codes with lots of datasets. But the result is still empty. It really confused me. Could you please apply your method with some sample datasets for example? If you want some datasets, I can send them to you.
> 
> Thanks,
> Sincerely,
> Xinze

Did you modify training.py accordingly based on your own dataset? When I used the code before, my preliminary result look like image attached here.
<img width="245" alt="Screen Shot 2019-10-11 at 6 09 42 PM" src="https://user-images.githubusercontent.com/29592494/66689872-753e4780-ec52-11e9-909d-854f9522e87d.png">

**Comment by Zhenya97** [comment]
Thank you for your reply.
I have made changes to the 'training.py' when I was running it. If you could give specific examples, for example, which parts need to be modified, I will modify it according to your instructions. Also, could you please give me an example of running remapping.py and calculate_weights.py file? I have some trouble in running them. 
I think I am very close to running the codes successfully. After that, I will write a detailed memo to you so that other novice users like me can refer to it.

Thanks,
Sincerely,
Chris
